### PR TITLE
Adding support for attribute exclusion in data filtering

### PR DESCRIPTION
**Pull request by neelismail**

### CODE REVIEW

File: api/src/main/resources/api/service_openapi.yaml [hunk 1]
```diff
@@ -444,6 +446,23 @@ components:
       type: string
       enum: ["AND", "OR"]
 
+    UnaryFilter:
+      type: object
+      description: A filter function for operating on a single operand.
+      properties:
+        operands:
+          type: array
```
Comment (tjennison-google) [inline api/src/main/resources/api/service_openapi.yaml:454]: This shouldn't be an array, since unlike the array operators there is only a single operand (as your description above says).

**Review by tjennison-google — CHANGES_REQUESTED** [review]
Please add tests for the new filter as well.

File: api/src/main/resources/api/service_openapi.yaml (hunk 1 above)
Comment (neelismail) [reply api/src/main/resources/api/service_openapi.yaml]: Done.

File: ui/cypress/integration/tests.js [hunk 2]
```diff
@@ -8,7 +8,7 @@ describe("Basic tests", () => {
     cy.get("button[id=insert-cohort]").click();
     cy.get("#text").type("New Cohort");
     cy.get("button:Contains(Create)").click();
-    cy.get("button:Contains(Add Criteria)").click();
```
Comment (tjennison-google) [inline ui/cypress/integration/tests.js:11]: Add a brief test for excluded criteria here like the included criteria. Just a couple lines to make sure it's not completely broken.

File: ui/src/cohort.ts [hunk 3]
```diff
@@ -34,16 +34,26 @@ function generateFilter(
       getCriteriaPlugin(criteria).generateFilter(entityVar, false)
     )
     .filter((filter) => filter) as Array<tanagra.Filter>;
+
```
Comment (tjennison-google) [inline ui/src/cohort.ts:37]: Remove this line.

In general, avoid introducing unnecessary changes. They make it more difficult to merge and review history.

File: ui/src/overview.tsx [hunk 4]
```diff
@@ -38,30 +38,44 @@ export default function Overview() {
...
 }
 
+function ParticipantsSelector(props: { kind?: tanagra.GroupKindEnum }) {
+  const cohort = useCohort();
+
+  return (
+    <Grid item xs={1}>
+      <Typography variant="h4">
+        {props.kind === tanagra.GroupKindEnum.Included
+          ? "Included"
```
Comment (tjennison-google) [inline ui/src/overview.tsx:56]: Just use "Included Participants" and "Excluded Participants" directly. Yes, "Participants" is repeated but this is harder to read, harder to translate, and only saves a few characters.

File: ui/cypress/integration/tests.js (hunk 2 above)
Comment (neelismail) [reply ui/cypress/integration/tests.js:11]: Done.

File: api/src/main/java/bio/terra/tanagra/service/search/SqlVisitor.java [hunk 5]
```diff
@@ -155,6 +156,23 @@ private static String convert(ArrayFunction.Operator operator) {
       }
     }
 
+    @Override
+    public String visitUnaryFunction(UnaryFunction unaryFunction) {
+      String operatorDelimiter = String.format("%s ", convert(unaryFunction.operator()));
+      // e.g. NOT ((operator1) OR operator2))
+
+      return operatorDelimiter.concat("(").concat(unaryFunction.operand().accept(this)).concat(")");
```
[the remark this answers is not in the document]
Comment (neelismail) [reply api/src/main/java/bio/terra/tanagra/service/search/SqlVisitor.java]: Done

File: ui/src/overview.tsx (hunk 4 above)
Comment (neelismail) [reply ui/src/overview.tsx]: Done.

File: ui/src/cohort.ts (hunk 3 above)
Comment (neelismail) [reply ui/src/cohort.ts]: Done. Sorry about that!

File: api/src/main/java/bio/terra/tanagra/service/search/SqlVisitor.java [hunk 6]
```diff
@@ -155,6 +156,23 @@ private static String convert(ArrayFunction.Operator operator) {
       }
     }
 
+    @Override
+    public String visitUnaryFunction(UnaryFunction unaryFunction) {
+      String operatorDelimiter = String.format("%s ", convert(unaryFunction.operator()));
+      // e.g. NOT ((operator1) OR operator2))
```
[the remark this answers is not in the document]
Comment (neelismail) [reply api/src/main/java/bio/terra/tanagra/service/search/SqlVisitor.java]: Done.

**Review by tjennison-google — APPROVED** [review]
